docs: add CI documentation (#88)
gardenlinux/docs-ng: docs/contributing/documentation/ci-architecture.md

  - /contributing/documentation/writing_good_docs.md
  - /contributing/documentation/aggregation-architecture.md
  - /contributing/documentation/adding-repos.md
  - /contributing/documentation/working-locally.md
  - /contributing/documentation/ci-workflows-reference.md
  - /contributing/documentation/configuration.md
  - /contributing/documentation/technical.md
  - /contributing/documentation/testing.md
  - /contributing/documentation/vitepress-features.md
---

# Documentation CI Architecture

## Purpose

The documentation CI pipeline serves three guarantees:

1. **Quality** — every documentation change in any aggregated repository is
   validated for broken links, spelling errors, and inclusive language _before_
   it can influence the published site.
2. **Preview** — contributors and reviewers can inspect exactly how a pull
   request's documentation changes will look on the live site, via a Netlify
   deploy preview attached to an automated draft PR in `docs-ng`.
3. **Always latest docs** — once a PR of an agregrated repository (source PR) is merged, the `repos-config.json`
   commit lock for that repository is automatically updated so the published

>    site reflects the latest content without any manual intervention.

Reviewer: ```suggestion
1. **Quality**: every documentation change in any aggregated repository is
   validated for broken links, spelling errors, and inclusive language _before_
   it can influence the published site.
2. **Preview**: contributors and reviewers can inspect exactly how a pull
   request's documentation changes will look on the live site, via a Netlify
   deploy preview attached to an automated draft PR in `docs-ng`.
3. **Always latest docs**: once a PR of an aggregrated repository (source PR) is merged, the `repos-config.json`
   commit lock for that repository is automatically updated so the published
   site reflects the latest content without any manual intervention.
```

- The config validator workflow ([`check-pr-main.yml`](https://github.com/gardenlinux/docs-ng/blob/main/.github/workflows/check-pr-main.yml)), which guards `main`
  against non-production refs in [`repos-config.json`](https://github.com/gardenlinux/docs-ng/blob/main/repos-config.json).
- The [`repos-config.json`](https://github.com/gardenlinux/docs-ng/blob/main/repos-config.json) file that records which repo and which exact commit
  is currently published.

### Aggregated repositories

Each repository listed in `repos-config.json` contains `docs-check.yml` workflow that:

- reacts to pull request and push events on documentation-relevant paths,
- calls the reusable `docs-checks.yml` from `docs-ng`, passing the PR's
  branch and commit as override inputs, and
- on success (or on merge), dispatches a `repository_dispatch` event to
  `docs-ng` to trigger the automated PR workflow.

### `gardenlinux-docs-bot` GitHub App

A GitHub App with write access to both `docs-ng` and the aggregated repositories.
It is used to:

- generate short-lived tokens that allow cross-repository API calls without
  using a personal access token,

> - open and update draft PRs in `docs-ng`, and
> - post preview comments back on aggregated repo PRs.
> 

Reviewer: ```suggestion
A GitHub App with write access to both `docs-ng` and the aggregated repositories.
It's responsible for generating short-lived tokens that allow cross-repository API calls without using a personal access token, opening and updating draft PRs in `docs-ng` and posting comments back to the PRs in the aggregated repo which include links to the Netlify preview.

```

Required secrets (configured in each aggregated repository):

| Secret | Purpose |
|--------|---------|
| `DOCS_BOT_APP_ID` | Identifies the GitHub App |
| `DOCS_BOT_PRIVATE_KEY` | Signs the JWT used to obtain installation tokens |

### Netlify

`docs-ng` has a [Netlify](https://www.netlify.com/) integration that automatically creates a deploy preview
for every open PR. The `docs-pr.yml` workflow posts the preview URL as a
comment on the originating aggregated repo PR so reviewers do not have to navigate
to `docs-ng` manually.

> ## High-Level Flow

Reviewer: ```suggestion
## Build Flow
```

        D --> E["Job: linkcheck\n(lychee)"]
        D --> F["Job: spelling\n(codespell via make spelling)"]
        D --> G["Job: woke\n(inclusive language)"]
    end

    E & F & G -->|all pass| H["notify-docs-ng job\nin aggregated repo"]
    H -->|repository_dispatch\nevent: docs-pr| I

    subgraph docsng2["gardenlinux/docs-ng"]
        I["docs-pr.yml\ncreates or updates branch\nauto/<repo>/<pr-number>"]
        I --> J["Updates repos-config.json\ncommit lock"]
        J --> K["git push -f\nOpens / updates draft PR\nin docs-ng"]
        K --> L["Posts preview comment\non source repo PR"]
        K --> M["check-pr-main.yml\nvalidates repos-config.json\nref must be main or semver"]
        K --> N["Netlify deploy preview\nbuilt from draft PR branch"]
    end

    L --> O[Reviewer sees\npreview link on source PR]
    N --> O

    O --> P{Source PR merged?}
    P -->|yes| Q["docs-check.yml\nclosed+merged event\nskips quality checks\ndispatches event=merged"]
    Q --> R["docs-pr.yml\nupdates lock to merge commit\nsets ref to base branch\nun-drafts PR in docs-ng"]
    R --> S[Maintainer merges\ndocs-ng PR]
    S --> T[Published site updated]

    P -->|no, more commits| A
```

> ## Sequence Diagrams

Reviewer: ```suggestion
## CI Flow Scenarios
```

> ### Scenario A — PR opened or synchronized in an aggregated repo

Reviewer: ```suggestion
### **Scenario A**: PR opened or synchronized in an aggregated repo
```

    DocsNG->>DocsNG: aggregate job<br/>(aggregate.py with overrides)
    DocsNG->>DocsNG: linkcheck job (lychee)
    DocsNG->>DocsNG: spelling job (codespell)
    DocsNG->>DocsNG: woke job (inclusive language)
    DocsNG-->>SrcRepo: all checks pass ✓
    SrcRepo->>SrcRepo: notify-docs-ng job runs
    SrcRepo->>Bot: create GitHub App token<br/>(repos: docs-ng, aggregated repo)
    SrcRepo->>DocsNG: repository_dispatch<br/>event_type: docs-pr<br/>payload: {repo, pr_number, commit_sha, ref, event: pr_success}
    DocsNG->>DocsNG: docs-pr.yml triggered
    DocsNG->>DocsNG: create/update branch<br/>auto/<repo>/<pr_number>
    DocsNG->>DocsNG: update repos-config.json<br/>commit lock → override-commit
    DocsNG->>DocsNG: git push -f to branch
    DocsNG->>DocsNG: open/update draft PR in docs-ng<br/>(status: Draft)
    DocsNG->>DocsNG: check-pr-main.yml validates<br/>repos-config.json refs
    Netlify->>Netlify: build deploy preview<br/>from draft PR branch
    Bot->>SrcRepo: post/update comment on source PR<br/>📚 Preview PR link<br/>😎 Netlify deploy preview URL
    Dev-->>SrcRepo: sees preview link in PR comment
```

> ### Scenario B — Source PR merged

Reviewer: ```suggestion
### **Scenario B**: Source PR merged
```
